### ISSUE TITLE
A little feedback and suggestion

### ISSUE THREAD

**Issue by LucasArusiewicz** [derkork/godot-statecharts#72]
## Overview

Recently, I've extensively used and tested the plugin, creating prototypes and more. However, after validating a mechanic, I realized that I had used an inappropriate name for an event, violating best practices and clean code principles. I went ahead to rename the event, which proved to be a bit tedious. While changing the code was straightforward, manually updating event names in all states was necessary.

After completing the renaming process, everything continued to work normally. Nevertheless, I identified an opportunity to enhance this aspect by abstracting and simplifying it within the plugin. My suggestion is to introduce an option for event registration, allowing users to select registered events or enter. If an event is selected, it would only update the String value. In principle, this way would have a more user-friendly interface and would maintain compatibility with existing projects.

## Visual State Construction

I also asked myself whether it would be possible to create a visual aspect of the State construction stage. In my research into graphics, I came across [stately.ai](https://stately.ai/), a tool with an interface that I really liked. I wondered if the plugin currently has a level of abstraction that allows the implementation of an interface using Godot's GraphEdit (Note: I haven't delved into the plugin's source code yet).

## Comments and suggestions

- Introduce a step to register events, allowing users to select events in addition to manually entering their names.
- Would it be possible to create an interface inspired by stately.ai to build and visualize the StateChart?

I look forward to your feedback on these suggestions and appreciate your efforts in improving and maintaining the plugin.

**Comment by derkork** [comment]
Hi Lucas, thanks a lot for taking the time to write all this. I agree that the string events are not ideal but given we need to interact with it from code, it's about the only thing we can do here. I can think of a few things to improve handling here:

- the event field in transitions gets an auto-complete that provides a list of all event names currently in use within the state chart. this would minimize the chance for a typo and could also speed up editing. 
![image](https://github.com/derkork/godot-statecharts/assets/327257/18f0670d-65f9-45be-a64c-7d369e9c1fb5)
  i'm not sure if the godot UI can do this, I will have to investigate. Doing it this way would make the event registration implicit without adding extra steps to the workflow.
- a new tool is added that allows to rename events that are used within a state chart: ![Snapshot](https://github.com/derkork/godot-statecharts/assets/327257/255a4e04-0518-4c29-87b5-c9437a79b06a) 
  this would make it quick and safe to rename all occurences of an event in a state chart and the editor has all the information to safely do it. You'd still need to rename the code occurrences.

Now for  the visual editor. I totally get the appeal of it and I'm a big fan of visual editing - I made [OpenSCAD Graph Editor](https://github.com/derkork/openscad-graph-editor).  I still decided against doing it for a few reasons:

- Godot's graph edit does not provide the necessary functionality to do it. While we can do flat node graphs pretty fine, we currently have not anything that would allow us to represent compound/parallel states. Something like this could potentially be added but it would require a good amount of work, especially for layouting  the hierarchy when you move nodes in nested states around.
- In general editing of nested state hierarchies is not very efficient in a diagram. I built the state chart for the platformer frog in stately.ai and while it was reasonably quick to set it up, doing it with nodes was a lot quicker as I don't need to layout the graph by hand all the time.  In addition there is no real nice way to move a state  from an inner nest level to an outer level while keeping its transitions e.g. like this 
![image](https://github.com/derkork/godot-statecharts/assets/327257/85106537-8e56-4471-922e-5bf55947bcd0) In a node tree this is trivial, i can just drag around nodes to a new parent.
- Nodes are already a familiar concept for users of Godot as almost everything is a node in Godot
- Nodes allow you to use the built-in features for scene composition and inheritance to make your state charts - or parts of them - reusable basically for free.

Finally, the amount of code needed for graph editor would very likely exceed the amount of code needed for everything else (I've got some experience on that with the OpenSCAD Graph Editor). I feel, that this time can better be spent elsewhere (for example this event name refactoring you suggested), especially given that a graph editor would not even be a better editing experience. What are your thoughts on this?

**Comment by SirPigeonz** [comment]
My experience with many different state machines both in Godot and Unreal show that Graph view is not that usefull if they don't support features that Godot Scene system already supports (and usually they don't). Those would be instancing, hierarchy, API to manage nodes and their flow through scripts and serializableto file modules.

I realy like current approach that State Charts uses. It's compact, fast to edit, easy to reause, modular and easy to extend.

Only two things I miss is refactoring tool for events, or the possibility to use predefined structure containing events, that would also help with maintaining documentation for State Trees.
Second one is a tool to better visualize transitions relations. Unfortunately Scene system is only good to visualize hierarchy, signaling and events are better viewed in graph form. Maybe separate debugging tool based on graphs similar to the one in Beehave?

Thanks for a great add-on! 😁

**Comment by officiallyaninja** [comment]
What are your thoughts on using nodes instead of strings for events? 
You could have a new `Event` Node that would be a child of StateChart (or possible another new `Events` Node) that is used to send events.  
  
So instead of doing 
```gdscript
$StateChart.send_event("attacked") 
```
You could create an `Attacked` Event Node, and then do 
```gdscript
$StateChart/Attacked.send_event()
```
Or something like that.
And then the event field will allow you to select a node, and renaming it shouldnt cause any issues. (assuming you use export vars instead of directly naming the nodes using node paths, but that's a common issue with node paths all over godot) 
  
This also has the benefits of having all the events in the node tree, so you don't have to remember all rhe events that you've defined.

**Comment by derkork** [comment]
I have a feeling this would introduce additional problems. Right now the problem is that we may have a typo in our event names somewhere. If we replace this by nodes we get additional sources of problems:

```gdscript
@onready var _my_event:Event  = $Graph/MyEvent  
```

The node path is essentially also a `String` which may have a typo. In addition this will break if we move the event nodes around or rename them. We could also export nodes/nodepaths like this:

```gscript
@export var my_event:Event   
# or 
@export_nodepath("Event") var my_event:NodePath 
```

This would fix the typo problem and would also potentially be auto-fixed by Godot when you move nodes around, but you still may forget to select an event in your export (which will then produce a null problem) or you may even select an event from a different state chart than the one you are working with. 

Also this would be detrimental to the usability of the library, you now need to create extra nodes and reference them just to send an event to the state chart. Finally this would be a breaking change breaking all existing projects out there. So I think providing an explicit refactoring for the event names and giving some autocomplete to prevent typos is less invasive and will hopefully not introduce additional sources of errors.

**Comment by LucasArusiewicz** [comment]
Hello @derkork , thank you for the feedback.

I really liked the autocomplete idea for the event field, but I believe there's nothing ready in Godot to implement this UI. I explored Godot a bit and found a trick using VSCode to refactor event names. Since scenes are not binary, VSCode interprets them as code and finds the references correctly, so it's just a matter of replacing them with the new value. 

![VSCode screenshot](https://github.com/derkork/godot-statecharts/assets/13992129/2525de33-7cdf-4281-bd04-ea35aae6c182)

I like the idea of having the option to refactor event names through the plugin itself. After some thought, I think we can design it similar to Godot's group manager.

![Godot group manager screenshot](https://docs.godotengine.org/pt-br/4.x/_images/groups_group_editor_window.webp)

Keeping things as simple as possible seems interesting to me. Representing events as just a string seems like a good option, essentially making it a tag. This simplifies the use of the plugin, similar to Godot groups. Godot has the Group Manager, which has an interface that displays indexed values (nodes present or not in a group) and facilitates batch editing.

Bringing this to the context of the plugin, it could be a list of registered events, transitions related to the event (perhaps even an option to add/link a new transition), and scripts associated with the event. From what I've seen, fetching and editing this information seems straightforward. The idea is to have something that facilitates batch editing of events and possibly expand to other plugin operations in the future.

### Graphical Editing

Regarding graphical editing, I hadn't paid attention to the details of data representation. What I had in mind was a visual representation of the tree. A correct visual representation of NinjaFrog would look like this: 

![Diagram screenshot](https://github.com/derkork/godot-statecharts/assets/13992129/53d73c11-050d-4036-a695-58778a427e38)


While practice and a well-defined approach may make it faster to create through nodes, having a visual representation could make it easier for some users. In more complex scenarios, visualizing and editing through Godot's node tree can become challenging. I moved almost all conditions and states of the player to the StateChart in one of my projects. The code became clean, but in the tree, I have 73 nodes, including CompoundState, AtomicState, and Transition. Collapsing the editor and looking at one at a time is fine, but understanding the overall context and how states relate would be easier in a graph.

Is there perhaps something already available in the Godot ecosystem? Like a tree editor – the Beehave debug seems somewhat like that, but I haven't tested it.

I understand that the visual aspect is extensive to program, but I believe we can define how it could be included in the plugin and identify the steps needed to complete it. As more people express interest in this feature, they could contribute to the plugin's development.

I am someone who, in my spare time, could easily program one thing or another and over time it would end up being ready, but if there were more people willing to help it would be great 😁

**Comment by derkork** [comment]
Autocomplete may actually be possible. There is nothing built in, but with a bit of event handling and custom drawing it might just work... 

https://github.com/derkork/godot-statecharts/assets/327257/d280f573-95e8-4f18-b6ff-fb2a0d5f5426

Still need the selection but looks definitely doable.

**Comment by derkork** [comment]
Okay, so the autocomplete idea didn't materialize after all, the issue is that while I can totally draw a completion popup, I cannot really detect the mouse on it. So I settled for a popup, which basically does the same thing and since it also supports typing it's almost a typeahead. I also added a facility to rename events together with full undo/redo support.

https://github.com/derkork/godot-statecharts/assets/327257/1d69a615-f9af-4cc2-ba60-1018cafcd688

I think this is already a nice improvement over the current status.

**Comment by officiallyaninja** [comment]
I prefer this over (just) a simple autocomplete imo being able to view and edit the events in the popup is a huge improvement

**Comment by derkork** [comment]
Released as 0.12.0. As for the visualization, I totally like the idea. However auto-layouting graphs is far from a solved problem. Most graph layouting tools I have seen produce abysmal results and not for a lack of trying. I actually have only seen one implementation that produces half-way decent results (yFiles) and this is both exceptionally expensive (licenses start at 22k€)  and not compatible with Godot. So any ideas how a graph layouter could actually work?

**Comment by devol-170** [comment]
Hi @derkork,
what tools did you try out for the graph layouting? According to statecharts.dev SCXML has been standardized over 10 years specifically for state charts.

Personally I like to work with plantUML, which supports multiple layout engines Graphviz (default), Smetana, VizJs and ELK see [here](https://plantuml.com/layout-engines)

The State Chart nodes and guards seem to adapt pretty well to plantUML's state diagram:
Atomic State -> Simple State, Compound State -> Composite State, Parallel State -> Concurrent State, Guard -> Conditional, History->History. Here's the state-diagram [documentation]( https://plantuml.com/state-diagram)

The downside of PlantUML is that don't see a straightforward option to run the image creation in engine. [Here](https://plantuml.com/running) are the integration options.

Tools like  "PlantUML Gizmo" Google Docs addon uses the [PlantUML Online Server ](https://www.plantuml.com/plantuml/uml/SyfFKj2rKt3CoKnELR1Io4ZDoSa700003) under the hood. Using a webservice  would enable a full "in editor" user/developer experience, but I'm not a fan of sending the layout to a 3rd party service. 

Thanks for the addon and kind regards!

**Comment by derkork** [comment]
Yes I'm aware of these tools, but as you have said getting these to work in engine is pretty much not possible without spending a disproportionate amount of effort. Using web services would invite all kinds of problems as well - the most immediate would be privacy concerns and the fact that these services usually cost money to operate and usually the providers of such services are not happy if people freeload on them. A middle ground could be a "copy plantuml to clipboard" option which generates a plantuml graph which one can then feed either into the server or a local copy of PlantUML. 

Also even with PlantUML the quality of the produced diagrams is rather sketchy, e.g. this is what i got from trying to recreate the OP's diagram:

```plantuml
@startuml

[*] --> Root

state Root {
    state Movement {
        state Grounded
        state Airborne {
            [*] --> CoyoteJumpEnabled
            state CoyoteJumpEnabled
            state DoubleJumpEnabled
            state CannotJump
            
            CoyoteJumpEnabled --> DoubleJumpEnabled: On Jump
            DoubleJumpEnabled --> CannotJump: On Jump
            CoyoteJumpEnabled --> CannotJump: after 0.5s 
        }
        
        Grounded --> Airborne: On Airborne
        Grounded --> DoubleJumpEnabled: On Jump
        
    }
    ---
    state AnimationControl {
        state Idle
        state Move
        
        Idle --> Move: On Moving
        Move --> Idle: On Idle
    }
}

@enduml
```

![Image](https://github.com/user-attachments/assets/d491387d-cda8-4e53-b67a-fd572c077181)

Its quite evident that the layout is a lot worse and this is only a rather simple case.

**Comment by devol-170** [comment]
I would really appreciate a single-file "copy PlantUML to clipboard" or "PUML export" feature. It might not be the prettiest visual representation, but it really helps with communication. The "out-of-editor" experience is probably not what the OPs of #72 and #162 had in mind, though.

I would advise against manually prettifying the layout.
> It’s true that when a diagram is big (or very big), manual placement could be useful. However, unfortunately, this goes against the PlantUML concept: https://crashedmind.github.io/PlantUMLHitchhikersGuide/layout/layout.html

Off-tangent:
I was curious if it's possible to "[zoom out](https://statecharts.dev/benefit-scales-with-complexity.html)" to a higher-level state chart. Sadly, that's not possible within a single file.

You can include modular statecharts, but that's also non-trivial. I took the .puml file for the OP's diagram as an example and extracted the substates into separate files.
[Here's](https://github.com/devol-170/plantuml-zoom-example) what it came down to.

**Comment by derkork** [comment]
Mermaids [render](https://mermaid.live/edit#pako:eNqNU01PwzAM_StRjmhFXLj0gDRtCIE0IcENwiFbvC6itas0QZqm_neSlDbduonl5I-X9-zEPvANKeA5b6y0sNSyMLISKPDz5otl2QN7I7LBj_nosYNA5k8XWdEPVIBDNGWeDDlUoE7jc23WZBDGN8LpFRe0JwsvrqofUa7LMUEiuQq0JOdT_zFJRLIBdJw99iZ6sdSJQM5ekU25Jriu0UH5wrXzouNrcmvBsLvb-2ZUb5vMZPW_ESn6L4i6vXMBe02PnfWnm2Xj-ZijrqTVhAtCa6iczsmzKuE0FqbqXBcBG6sKgFiINzQWCRESERGgEZH4fYEtn_EKTCW18kMfaxHc7vwIC557U0nzLXiHk87S-x43PLfGwYz7Vyl2PN_KsvGeq1XamCFaS_wgSj4obcmsuh2Lq9b-AhwnCRg) is actually much more usable:

![Image](https://github.com/user-attachments/assets/70c62ba6-c64d-4431-ab47-ce79b1c47e25)

Funny enough they use the same syntax as PlantUML. I also found [Kroki](https://kroki.io/) which is a public service for rendering all kinds of text-based diagrams - they explicitly state that it is free to use for everybody. So how about this:

- We add an option put the chart as PlantUML diagram to the clipboard.
- We add a way to send this directly to Kroki and show the image directly in the editor (this needs to be enabled manually for privacy reasons and with the caveat that the service might disappear).

**Comment by derkork** [comment]
I was just made aware that Mermaid has problems [rendering descriptions](https://github.com/chickensoft-games/LogicBlocks/issues/35) of states which would be sort of important, so I guess I'll stick to PlantUML for now even if the diagrams look less nice.

**Comment by derkork** [comment]
So well, i think i got the basic transformation working. Some stuff looks pretty neat:


![Image](https://github.com/user-attachments/assets/919af15b-a428-4216-89e5-e0686d3dfe2d)


Some stuff looks less neat (notice the overlapping labels):


![Image](https://github.com/user-attachments/assets/f736ecf1-d859-4072-99a3-0a4f6e7597b2)


I'm not sure how much I can do about this, but i guess _some_ visualization is better than none.

**Comment by devol-170** [comment]
There are several [overlapping](https://forum.plantuml.net/tag/overlapping) issues with PlantUML. That's a concern of the rendering tool. You also don't know which tool the user will use. Some charts will look broken and ugly depending on their complexity.
You should keep the graph as basic as possible and without layout tweaking in my opinion.

Really strange, that it managed to produce a bidirectional arrow though.

**Comment by derkork** [comment]
I agree, I will not get down the rabbit hole of trying to auto-tweak the graph layout, that is just a never-ending game of whack-a-mole.  For the "bidirectionaly arrow" these are actually two distinct transitions, they just render on top of each other for some reason.

**Comment by derkork** [comment]
We're getting somewhere:

https://github.com/user-attachments/assets/a4b40397-6de9-45f9-a992-39e7e5a843d5

Still needs a shitload of polishing and error handling, but we're getting somewhere.

**Comment by derkork** [comment]
When following the editor theme this looks a lot less alien:

![Image](https://github.com/user-attachments/assets/cc4e4a77-aaeb-4ba9-ab0c-ea0721047a74)

![Image](https://github.com/user-attachments/assets/dc7f2356-d8c1-4b4a-9a8a-736d3a4d7db9)

**Comment by derkork** [comment]
>  will it be possible to click on the graph element (a state node or transition edge) to jump to the corresponding node in the scene tree and inspector?

No, this is is just a PNG image.

**Comment by derkork** [comment]
SVG would allow me to read in some data and maybe trace the shape back to the node from which it originated but unfortunately Godot doesn't render text in SVGs, so I just get a lot of empty boxes when using SVG. This is turning into a real rabbit hole.
